src/boot/efi/boot.c

```diff
@@ -1951,6 +1961,95 @@ static VOID config_entry_add_linux(
         uefi_call_wrapper(linux_dir->Close, 1, linux_dir);
 }
 
+#ifndef DEVICE_TREE_GUID
+#define DEVICE_TREE_GUID \
+        { 0xb1b621d5, 0xf19c, 0x41a5, {0x83, 0x0b, 0xd9, 0x15, 0x2c, 0x69, 0xaa, 0xe0} }
+#endif
+static const EFI_GUID fdt_guid = DEVICE_TREE_GUID;
+
+static EFI_STATUS devicetree_get(const void **fdt) {
+        EFI_CONFIGURATION_TABLE *entry = ST->ConfigurationTable;
+        EFI_CONFIGURATION_TABLE *end = entry + ST->NumberOfTableEntries;
+
+        for (; entry < end; entry++) {
+                if (CompareGuid(&entry->VendorGuid, (EFI_GUID *)&fdt_guid) == 0) {
+                        *fdt = entry->VendorTable;
+                        return EFI_SUCCESS;
+                }
+        }
+
+        return EFI_UNSUPPORTED;
+}
+
+static const void *devicetree_orig;
+static EFI_PHYSICAL_ADDRESS devicetree_addr;
+static UINTN devicetree_pages;
+
+static EFI_STATUS devicetree_install(EFI_FILE_HANDLE root_dir, const CHAR16 *name) {
+        _cleanup_(FileHandleClosep) EFI_FILE_HANDLE handle = NULL;
+        _cleanup_freepool_ EFI_FILE_INFO *info = NULL;
+        UINTN size;
+        EFI_STATUS err;
+        void *fdt;
+
+        err = devicetree_get(&devicetree_orig);
+        if (EFI_ERROR(err))
+                return err;
+
+        err = uefi_call_wrapper(root_dir->Open, 5, root_dir, &handle, (CHAR16 *)name,
+                        EFI_FILE_MODE_READ, EFI_FILE_READ_ONLY);
+        if (EFI_ERROR(err))
+                return err;
+
+        info = LibFileInfo(handle);
+        if (!info)
+                return EFI_OUT_OF_RESOURCES;
+        if (info->FileSize < FDT_V1_SIZE || info->FileSize > 32 * 1024 * 1024)
+                /* 32MB device tree blob doesn't seem right */
+                return EFI_INVALID_PARAMETER;
+
+        size  = info->FileSize;
+        FreePool(TAKE_PTR(info));
```

Comment:
unnecessary, no? you already free it via `_cleanup_freepool_` anyway, so no need to do this manually here

Reply:
Right, it'll just do it a bit earlier. I'll remove